mm2src/mm2_main/tests/docker_tests/z_coin_docker_tests.rs

```diff
@@ -0,0 +1,242 @@
+use bitcrypto::dhash160;
+use coins::z_coin::{z_coin_from_conf_and_params_with_docker, z_send_dex_fee, ZCoin, ZcoinActivationParams,
+                    ZcoinRpcMode};
+use coins::{coin_errors::ValidatePaymentError, CoinProtocol, DexFee, PrivKeyBuildPolicy, RefundPaymentArgs,
+            SendPaymentArgs, SpendPaymentArgs, SwapOps, SwapTxTypeWithSecretHash, ValidateFeeArgs};
+use common::{executor::Timer, now_sec};
+use lazy_static::lazy_static;
+use mm2_core::mm_ctx::{MmArc, MmCtxBuilder};
+use mm2_number::MmNumber;
+use mm2_test_helpers::for_tests::{new_mm2_temp_folder_path, zombie_conf};
+use rand::distributions::Alphanumeric;
+use rand::{thread_rng, Rng};
+use tokio::sync::Mutex;
+
+// https://github.com/KomodoPlatform/librustzcash/blob/4e030a0f44cc17f100bf5f019563be25c5b8755f/zcash_client_backend/src/data_api/wallet.rs#L72-L73
+lazy_static! {
+    static ref TEST_MUTEX: Mutex<()> = Mutex::new(());
+}
+
+/// Build asset `ZCoin` from ticker and spendingkey str without filling the balance.
+pub async fn z_coin_from_spending_key(spending_key: &str) -> (MmArc, ZCoin) {
+    let ctx = MmCtxBuilder::new().into_mm_arc();
+    let mut conf = zombie_conf();
+    let params = ZcoinActivationParams {
+        mode: ZcoinRpcMode::Native,
+        ..Default::default()
+    };
+    let pk_data = [1; 32];
+    let salt: String = thread_rng()
+        .sample_iter(&Alphanumeric)
+        .take(4)
+        .map(char::from)
+        .collect();
+    let db_folder = new_mm2_temp_folder_path(None).join(format!("ZOMBIE_DB_{}", salt));
+    std::fs::create_dir_all(&db_folder).unwrap();
+    let protocol_info = match serde_json::from_value::<CoinProtocol>(conf["protocol"].take()).unwrap() {
+        CoinProtocol::ZHTLC(protocol_info) => protocol_info,
+        other_protocol => panic!("Failed to get protocol from config: {:?}", other_protocol),
+    };
+
+    let coin = z_coin_from_conf_and_params_with_docker(
+        &ctx,
+        "ZOMBIE",
+        &conf,
+        &params,
+        PrivKeyBuildPolicy::IguanaPrivKey(pk_data.into()),
+        db_folder,
+        protocol_info,
+        spending_key,
+    )
+    .await
+    .unwrap();
+
+    (ctx, coin)
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn zombie_coin_send_and_refund_maker_payment() {
+    let _lock = TEST_MUTEX.lock().await;
```

Comment:
I think we should set this lock on the API level (not in tests only)
(to prevent [this issue with parallel calls](https://github.com/KomodoPlatform/librustzcash/blob/4e030a0f44cc17f100bf5f019563be25c5b8755f/zcash_client_backend/src/data_api/wallet.rs#L72-L73), right?

Reply:
no sure how we can approach this btw.. since we need to wait from start to finish of a pirate swap.

Reply:
I think we just should prevent a note to be selected the second time from the received_notes table.
It should not be selected after we update its `spent` status with the store_sent_tx fn.
So I guess, we need to set a lock before where we add notes and release it after the store_sent_tx fn called. The places in code where we add notes are: init_withdraw, z_send_dex_fee, z_send_htlc

Reply:
I made a [sample code](https://github.com/KomodoPlatform/komodo-defi-framework/compare/fix-zombie-tests...dimxy:komodo-defi-framework:fix-zombie-tests?expand=1#) with Z_NOTE_LOCK in the send_outputs fn (it's not ready for prod, just to prove the idea of a lock to prevent access to the available notes)

I also had to add a simple wait_for_z_balance fn, otherwise I had insufficient balance error. So probably the  spendable_notes_required_for_tx fn does work properly when different ZCoin {} created in concurrent zombie tests or its confirmation logic is not sufficient: I noticed that transactions may stay for few blocks in mempool so I wait for 5 blocks in wait_for_z_balance.